feat: add C ndarray API and refactor `blas/ext/base/ssort2ins`

0PrashantYadav0 opened the pull request:
---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes. report:
  - task: lint_filenames status: passed
  - task: lint_editorconfig status: passed
  - task: lint_markdown status: na
  - task: lint_package_json status: na
  - task: lint_repl_help status: passed
  - task: lint_javascript_src status: passed
  - task: lint_javascript_cli status: na
  - task: lint_javascript_examples status: passed
  - task: lint_javascript_tests status: na
  - task: lint_javascript_benchmarks status: na
  - task: lint_python status: na
  - task: lint_r status: na
  - task: lint_c_src status: passed
  - task: lint_c_examples status: passed
  - task: lint_c_benchmarks status: passed
  - task: lint_c_tests_fixtures status: na
  - task: lint_shell status: na
  - task: lint_typescript_declarations status: passed
  - task: lint_typescript_tests status: na
  - task: lint_license_headers status: passed ---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes. report:
  - task: run_javascript_examples status: na
  - task: run_c_examples status: na
  - task: run_cpp_examples status: na
  - task: run_javascript_readme_examples status: na
  - task: run_c_benchmarks status: na
  - task: run_cpp_benchmarks status: na
  - task: run_fortran_benchmarks status: na
  - task: run_javascript_benchmarks status: na
  - task: run_julia_benchmarks status: na
  - task: run_python_benchmarks status: na
  - task: run_r_benchmarks status: na
  - task: run_javascript_tests status: na ---

Resolves None.

## Description

> What is the purpose of this pull request?

This pull request:

-   add C `ndarray` API and refactor/cleanup `blas/ext/base/ssort2ins`

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves None.

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

0PrashantYadav0 committed:
feat: add C ndarray interface and refactor implementation

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: passed
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

0PrashantYadav0 committed:
docs: update README for ssort2ins function and add C API documentation

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: passed
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/ssort2ins/README.md:
```diff
@@ -54,9 +54,9 @@ The function has the following parameters:
--   **strideY**: `y` index increment.
+-   **strideY**: stride length for `y`.
 
 The `N` and `stride` parameters determine which elements in the strided arrays are accessed at runtime. For example, to sort every other element
```
```suggestion
The `N` and stride parameters determine which elements in the strided arrays are accessed at runtime. For example, to sort every other element:
```

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/ssort2ins/README.md:
```diff
@@ -207,6 +193,137 @@ console.log( y );
+
+-   **N**: `[in] CBLAS_INT` number of indexed elements.
+-   **order**: `[in] float` sort order. If `order < 0.0`, the input strided array `x` is sorted in **decreasing** order. If `order > 0.0`, the input strided array `x` is sorted in **increasing** order. If `order == 0.0`, the input strided arrays are left unchanged.
+-   **X**: `[inout] float*` input array.
```
```suggestion
-   **X**: `[inout] float*` first input array.
```

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/ssort2ins/README.md:
```diff
@@ -207,6 +193,137 @@ console.log( y );
+-   **order**: `[in] float` sort order. If `order < 0.0`, the input strided array `x` is sorted in **decreasing** order. If `order > 0.0`, the input strided array `x` is sorted in **increasing** order. If `order == 0.0`, the input strided arrays are left unchanged.
+-   **X**: `[inout] float*` input array.
+-   **strideX**: `[in] CBLAS_INT` stride length for `X`.
+-   **Y**: `[inout] float*` input array.
```
```suggestion
-   **Y**: `[inout] float*` second input array.
```

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/ssort2ins/README.md:
```diff
@@ -207,6 +193,137 @@ console.log( y );
+
+-   **N**: `[in] CBLAS_INT` number of indexed elements.
+-   **order**: `[in] float` sort order.
+-   **X**: `[inout] float*` input array.
```
```suggestion
-   **X**: `[inout] float*` first input array.
```

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/ssort2ins/README.md:
```diff
@@ -207,6 +193,137 @@ console.log( y );
+-   **X**: `[inout] float*` input array.
+-   **strideX**: `[in] CBLAS_INT` stride length for `X`.
+-   **offsetX**: `[in] CBLAS_INT` starting index for `X`.
+-   **Y**: `[inout] float*` input array.
```
```suggestion
-   **Y**: `[inout] float*` second input array.
```

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/ssort2ins/README.md:
```diff
@@ -207,6 +193,137 @@ console.log( y );
+    // Print the result:
+    for ( int i = 0; i < 8; i++ ) {
+        printf( "x[ %i ] = %f\n", i, x[ i ] );
+        printf( "y[ %i ] = %lf\n", i, y[ i ] );
```
```suggestion
        printf( "y[ %i ] = %f\n", i, y[ i ] );
```

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/ssort2ins/benchmark/c/benchmark.unsorted_random.length.c:
```diff
@@ -0,0 +1,203 @@
+	float *y;
+	int i;
+
+	x = (float *)malloc( len * sizeof(float) );
+	y = (float *)malloc( len * sizeof(float) );
```
```suggestion
	x = ( float * )malloc( len * sizeof(float) );
	y = ( float * )malloc( len * sizeof(float) );
```

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/ssort2ins/docs/repl.txt:
```diff
@@ -113,7 +113,7 @@
+        Stride length for `x`.
 
     offsetX: integer
         Starting index of `x`.
```
```suggestion
        Starting index for `x`.
```

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/ssort2ins/docs/repl.txt:
```diff
@@ -122,7 +122,7 @@
+        Stride length for `y`.
 
     offsetY: integer
         Starting index of `y`.
```
```suggestion
        Starting index for `y`.
```

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/ssort2ins/examples/c/example.c:
```diff
@@ -33,12 +32,12 @@ int main( void ) {
 	for ( int i = 0; i < 8; i++ ) {
 		printf( "x[ %i ] = %f\n", i, x[ i ] );
-		printf( "y[ %i ] = %"PRId64"\n", i, (int64_t)y[ i ] );
+		printf( "y[ %i ] = %lf\n", i, y[ i ] );
```
```suggestion
		printf( "y[ %i ] = %f\n", i, y[ i ] );
```

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/ssort2ins/src/addon.c:
```diff
@@ -1,7 +1,7 @@
 * @license Apache-2.0
 *
-* Copyright (c) 2020 The Stdlib Authors.
+* Copyright (c) 2018 The Stdlib Authors.
```
```suggestion
* Copyright (c) 2020 The Stdlib Authors.
```

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/ssort2ins/src/main.c:
```diff
@@ -28,22 +29,43 @@
 * @param order    sort order
 * @param X        first input array
-* @param strideX  `X` index increment
+* @param strideX  stride length for `x`
```
```suggestion
* @param strideX  stride length for `X`
```

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/ssort2ins/src/main.c:
```diff
@@ -28,22 +29,43 @@
+* @param strideX  stride length for `x`
 * @param Y        second input array
-* @param strideY  `Y` index increment
+* @param strideY  stride length for `y`
```
```suggestion
* @param strideY  stride length for `Y`
```

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/ssort2ins/src/main.c:
```diff
@@ -28,22 +29,43 @@
+/**
+* Simultaneously sorts two signle-precision floating-point strided arrays based on the sort order of the first array using insertion sort and alternative indexing semantics.
+*
+* @param N        number of indexed elements
+* @param order    sort order
+* @param X        first input array
+* @param strideX  stride length for `x`
+* @param offsetX  starting index for `x`
+* @param Y        second input array
+* @param strideY  stride length for `y`
+* @param offsetY  starting index for `y`
```
```suggestion
* @param N        number of indexed elements
* @param order    sort order
* @param X        first input array
* @param strideX  stride length for `X`
* @param offsetX  starting index for `X`
* @param Y        second input array
* @param strideY  stride length for `Y`
* @param offsetY  starting index for `Y`
```

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/ssort2ins/src/main.c:
```diff
@@ -54,26 +76,26 @@ void c_ssort2ins( const int64_t N, const float order, float *X, const int64_t st
-		ly = (N-1) * sy;
-	}
+	fx = ox;              // first index
+	lx = fx + ((N-1)*sx); // last index
```
```suggestion
	lx = fx + ( (N-1)*sx ); // last index
```

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/ssort2ins/src/main.c:
```diff
@@ -54,26 +76,26 @@ void c_ssort2ins( const int64_t N, const float order, float *X, const int64_t st
+	ix = fx + sx;
+
+	fy = oy;              // first index
+	ly = fy + ((N-1)*sy); // last index
```
```suggestion
	ly = fy + ( (N-1)*sy ); // last index
```

headlessNode reviewed (commented):
Thanks for working on this. In addition to these suggestions you also need to update the line `100` in the `repl`. It should be same as in the `README.md`

0PrashantYadav0 committed:
docs: improve clarity in README and documentation for ssort2ins function

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: passed
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

0PrashantYadav0 commented:
Thank you @headlessNode for review. I made the changes you asked for. Please review again.

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/ssort2ins/benchmark/c/benchmark.unsorted_random.length.c:
```diff
@@ -0,0 +1,203 @@
+	float *y;
+	int i;
+
+	x = ( float * )malloc( len * sizeof(float) );
+	y = ( float * )malloc( len * sizeof(float) );
```
```suggestion
	x = (float *)malloc( len * sizeof(float) );
	y = (float *)malloc( len * sizeof(float) );
```

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/ssort2ins/benchmark/c/benchmark.unsorted_random.length.c:
```diff
@@ -0,0 +1,203 @@
+	float *y;
+	int i;
+
+	x = ( float * )malloc( len * sizeof(float) );
+	y = ( float * )malloc( len * sizeof(float) );
```
```suggestion
	x = (float *)malloc( len * sizeof(float) );
	y = (float *)malloc( len * sizeof(float) );
```

headlessNode committed:
docs: update offset description

Signed-off-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>

headlessNode reviewed (changes requested):
Just found out, our convention is to NOT add spacing around parens when casting. This applies to other similar PRs as well.

0PrashantYadav0 committed:
fix: improve code formatting in benchmark C files

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/ssort2ins/README.md:
```diff
@@ -207,6 +193,137 @@ console.log( y );
+The function accepts the following arguments:
+
+-   **N**: `[in] CBLAS_INT` number of indexed elements.
+-   **order**: `[in] float` sort order.
```
```suggestion
-   **order**: `[in] float` sort order. If `order < 0.0`, the input strided array `x` is sorted in **decreasing** order. If `order > 0.0`, the input strided array `x` is sorted in **increasing** order. If `order == 0.0`, the input strided arrays are left unchanged.
```

headlessNode committed:
docs: apply suggestions from code review

Signed-off-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>

headlessNode commented:
LGTM.

0PrashantYadav0 commented:
Thanks for approval.

0PrashantYadav0 commented:
@kgryte Please review and merge this pr. Thank you.
